components/tidb_query_common/src/metrics.rs

```diff
@@ -29,16 +30,45 @@ make_auto_flush_static_metric! {
     }
 }
 
-lazy_static::lazy_static! {
+make_static_metric! {
+    pub label_enum MemTrace {
+        aggr_fast_hash,
+        aggr_slow_hash,
+        aggr_stream,
+        aggr_simple,
+        top_n,
+        index_scan,
+        table_scan,
+        limit,
+        projection,
+        selection,
+        chunk,
+    }
+
+    pub struct MemTraceGauge : IntGauge {
+        "type" => MemTrace,
+    }
+}
+
+lazy_static! {
+    pub static ref MEMTRACE_QUERY_EXECUTOR: MemTraceGauge = register_static_int_gauge_vec!(
+        MemTraceGauge,
+        "tikv_query_executor_memory_usage",
```

Comment:
why not record the memory usage in current memory trace metrics

Reply:
I did go down that path first, the reason I moved away from it are:

- There is no hierarchy, it's a single level counter
- Without any clear benefit it seemed like over engineering

In case I've missed or misunderstood something, can you please elaborate.

Reply:
It's hierarchical by `sub_trace`, check `MEMTRACE_ANALYZE` for reference. I think we can make `index_scan/table_scan/...` be the sub-trace for `MEMTRACE_QUERY` which is a sub-trace to `MEMTRACE_ROOT` of coprocessor mod